tests/ga/test_agent_update.py

```diff
@@ -0,0 +1,575 @@
+
+            with patch("azurelinuxagent.common.conf.get_autoupdate_enabled", return_value=autoupdate_enabled):
+                with patch("azurelinuxagent.common.conf.get_autoupdate_frequency", return_value=autoupdate_frequency):
+                    with patch("azurelinuxagent.common.conf.get_autoupdate_gafamily", return_value="Prod"):
+                        with patch("azurelinuxagent.ga.agent_update.add_event") as mock_telemetry:
+                            agent_update_handler = get_agent_update_handler(protocol)
+                            agent_update_handler._protocol = protocol
+                            yield agent_update_handler, mock_telemetry
+
+    def __assert_agent_directories_available(self, versions):
+        for version in versions:
+            self.assertTrue(os.path.exists(self.agent_dir(version)), "Agent directory {0} not found".format(version))
+
+    def __assert_agent_directories_exist_and_others_dont_exist(self, versions):
+        self.__assert_agent_directories_available(versions=versions)
+        other_agents = [agent_dir for agent_dir in self.agent_dirs() if
+                        agent_dir not in [self.agent_dir(version) for version in versions]]
+        self.assertFalse(any(other_agents),
+                         "All other agents should be purged from agent dir: {0}".format(other_agents))
+
+    def test_it_should_not_update_when_autoupdate_disabled(self):
+        self.prepare_agents(count=1)
+        with self.__get_agent_update_handler(autoupdate_enabled=False) as (agent_update_handler, mock_telemetry):
+            agent_update_handler.run()
+            self.__assert_agent_directories_exist_and_others_dont_exist(versions=[str(CURRENT_VERSION)])
+            self.assertEqual(0, len([kwarg['message'] for _, kwarg in mock_telemetry.call_args_list if
+                                     "Found requested version" in kwarg['message'] and kwarg[
+                                         'op'] == WALAEventOperation.AgentUpgrade]), "Unwanted upgrade")
+
+    def test_it_should_not_agent_update_if_last_attempted_update_time_not_elapsed(self):
+        self.prepare_agents(count=1)
+        data_file = DATA_FILE.copy()
+        data_file["ext_conf"] = "wire/ext_conf_requested_version.xml"
+        version = "5.2.0.1"
+        with self.__get_agent_update_handler(test_data=data_file, autoupdate_frequency=10) as (agent_update_handler, mock_telemetry):
+            with patch.object(conf, "get_enable_ga_versioning", return_value=True):
+                agent_update_handler._protocol.mock_wire_data.set_extension_config_requested_version(version)
+                agent_update_handler._protocol.mock_wire_data.set_incarnation(2)
+                agent_update_handler._protocol.update_goal_state()
+                agent_update_handler.run()
+
+                self.assertEqual(1, len([kwarg['message'] for _, kwarg in mock_telemetry.call_args_list if
```

Comment:
This search is repeated several times with different messages. Let's capture it in a helper function to improve readability of the tests

[kwarg['message'] for _, kwarg in mock_telemetry.call_args_list if
                                         "Found requested version: {0}".format(version) in kwarg[
                                             'message'] and kwarg[
                                             'op'] == WALAEventOperation.AgentUpgrade]

Reply:
Also, consider checking whether the agent was updated or not by other means that telemetry messages. Using messages can break if we change the message in the agent, and some test may even fail silently.

Reply:
> This search is repeated several times with different messages. Let's capture it in a helper function to improve readability of the tests
> 
> [kwarg['message'] for _, kwarg in mock_telemetry.call_args_list if "Found requested version: {0}".format(version) in kwarg[ 'message'] and kwarg[ 'op'] == WALAEventOperation.AgentUpgrade]

yes, I realized later, doing it in next commit

Reply:
> Also, consider checking whether the agent was updated or not by other means that telemetry messages. Using messages can break if we change the message in the agent, and some test may even fail silently.

I believe tests are meant to break if we introduce something which are not expecting. Telemetry check I added is additional checks along with others.

`some test may even fail silently.`
I don't understand this.

Reply:
If we update the telemetry message, the asserts that check for the presence of the message (self.assertEqual(1, len(upgrade_event_msgs)) will fail on those are easy to fix.

However, the tests that check for the absence of the message (len == 0) will continue to pass, event if we introduce a bug .